File: src/main/java/LeetCode/TaskScheduler.java
package LeetCode;

import java.util.*;

public class TaskScheduler {
    public int leastInterval(char[] tasks, int n) {
        if (n == 0) return tasks.length;
        Map<Character, Integer> map = new HashMap<>();
        for (char c : tasks) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        maxHeap.addAll(map.values());
        int cycles = 0;
        while (!maxHeap.isEmpty()) {
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                if (!maxHeap.isEmpty()) {
                    temp.add(maxHeap.poll());
                }
            }
            for (int i : temp) {
                if (--i > 0) maxHeap.add(i);
            }
            cycles += maxHeap.isEmpty() ? temp.size() : n + 1;
        }
        return cycles;
    }
}

File: python/leetcode/task_scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        same_tasks = {}
        for c in tasks:
            if c in same_tasks:
                same_tasks[c] += 1
            else:
                same_tasks[c] = 1
        max_heap = []
        for i in same_tasks.values():
            heapq.heappush(max_heap, -i)
        cycles = 0
        while len(max_heap) != 0:
            temp = []
            for i in range(n+1):
                if len(max_heap) != 0:
                    temp.append(-heapq.heappop(max_heap))
            for i in temp:
                i -= 1
                if i > 0: heapq.heappush(max_heap, -i)
            if len(max_heap) == 0:
                cycles += len(temp)
            else:
                cycles += n + 1
        return cycles